src/ucp/core/ucp_ep.c

```diff
@@ -1287,6 +1306,31 @@ static void ucp_ep_discard_lanes_callback(void *request, ucs_status_t status,
     ucs_free(arg);
 }
 
+static ucs_status_t ucp_ep_failed_op(uct_ep_h ep)
+{
+    ucp_ep_failed_arg_t *arg = ucs_derived_of(ep, ucp_ep_failed_arg_t);
+
+    return arg->status;
+}
+
+static ssize_t ucp_ep_failed_bc_op(uct_ep_h ep)
+{
+    ucp_ep_failed_arg_t *arg = ucs_derived_of(ep, ucp_ep_failed_arg_t);
+
+    return arg->status;
+}
+
+static void ucp_ep_failed_destroy(uct_ep_h ep)
+{
+    ucp_ep_failed_arg_t *arg = ucs_derived_of(ep, ucp_ep_failed_arg_t);
+
+    if (!arg->free || (--arg->counter > 0)) {
```

Comment:
according to ucp_worker_discard_uct_ep_destroy_progress, we first destroy uct_ep and then release discard_arg, so should be able to place the failed ep on discard_arg and not use a separate counter